lib/stdlib/src/supervisor.erl

```diff
@@ -41,170 +54,230 @@
-		       period => pos_integer()}        % optional
-                   | {RestartStrategy :: strategy(),
-                      Intensity :: non_neg_integer(),
-                      Period :: pos_integer()}.
+-type worker() :: 'worker' | 'supervisor'.
+-type sup_name() ::
+    {'local', Name :: atom()} |
+    {'global', Name :: atom()} |
+    {'via', Module :: module(), Name :: any()}.
+
+-type sup_ref() ::
+    (Name :: atom()) |
+    {Name :: atom(), Node :: node()} |
+    {'global', Name :: atom()} |
+    {'via', Module :: module(), Name :: any()} |
+    pid().
+
+% mandatory
+-type child_spec() ::
+    #{
+        id := child_id(),
+        % mandatory
+        start := mfargs(),
+        % optional
+        restart => restart(),
+        % optional
+        shutdown => shutdown(),
+        % optional
+        type => worker(),
+        % optional
+        modules => modules()
+    } |
+    {Id :: child_id(), StartFunc :: mfargs(), Restart :: restart(), Shutdown :: shutdown(),
+        Type :: worker(), Modules :: modules()}.
+
+-type strategy() ::
+    'one_for_all' |
+    'one_for_one' |
+    'rest_for_one' |
+    'simple_one_for_one'.
+
+% optional
+-type sup_flags() ::
+    #{
+        strategy => strategy(),
+        % optional
+        intensity => non_neg_integer(),
+        % optional
+        period => pos_integer()
+    } |
+    {RestartStrategy :: strategy(), Intensity :: non_neg_integer(), Period :: pos_integer()}.
+
 -type children() :: {Ids :: [child_id()], Db :: #{child_id() => child_rec()}}.
 
 %%--------------------------------------------------------------------------
 %% Defaults
--define(default_flags, #{strategy  => one_for_one,
-			 intensity => 1,
-			 period    => 5}).
--define(default_child_spec, #{restart  => permanent,
-			      type     => worker}).
+-define(default_flags, #{
+    strategy => one_for_one,
+    intensity => 1,
+    period => 5
+}).
+
+-define(default_child_spec, #{
+    restart => permanent,
+    type => worker
+}).
+
 %% Default 'shutdown' is 5000 for workers and infinity for supervisors.
 %% Default 'modules' is [M], where M comes from the child's start {M,F,A}.
 
 %%--------------------------------------------------------------------------
 
--record(child, {% pid is undefined when child is not running
-	        pid = undefined :: child()
-	                         | {restarting, pid() | undefined}
-	                         | [pid()],
-		id              :: child_id(),
-		mfargs          :: mfargs(),
-		restart_type    :: restart(),
-		shutdown        :: shutdown(),
-		child_type      :: worker(),
-		modules = []    :: modules()}).
+% pid is undefined when child is not running
+-record(child, {
+    pid = undefined ::
+        child() |
+        {restarting, pid() | undefined} |
+        [pid()],
+    id :: child_id(),
+    mfargs :: mfargs(),
+    restart_type :: restart(),
+    shutdown :: shutdown(),
+    child_type :: worker(),
+    modules = [] :: modules()
+}).
+
```

Comment:
I believe this is easier to automate but the left side is easier to read.

Reply:
I have added the comment as a data point to another issue https://github.com/WhatsApp/erlfmt/issues/134

What part do you think is making the left side easier to read other wise?

Reply:
The alignment of "::" makes it easier to read!

Reply:
Oh wow yes, really good point.  This is a good example.

The reason we decided not to make the formatter do any alignment is because we do not want changing a single line to result in a big change.

For example if we take the example here:
```erlang
-record(child, {% pid is undefined when child is not running
             pid = undefined :: child()
                              | {restarting, pid() | undefined}
                              | [pid()],
             id              :: child_id(),
             mfargs          :: mfargs(),
             restart_type    :: restart(),
             shutdown        :: shutdown(),
             child_type      :: worker(),
             modules = []    :: modules()}).
```

If we provide `restart_type` with a default or add a new field, then we have to shift all types to keep them all aligned.

```erlang
-record(child, {% pid is undefined when child is not running
             pid = undefined :: child()
                              | {restarting, pid() | undefined}
                              | [pid()],
             id              :: child_id(),
             mfargs          :: mfargs(),
             restart_type    :: restart(),
             shutdown        :: shutdown(),
             child_type      :: worker(),
             new_longer_field :: longer_type(),
             modules = []    :: modules()}).
```

This would result in a modification of each line, which is something we would like to avoid.